FeLib/Source/specialkeys.cpp

```diff
@@ -0,0 +1,264 @@
+//REQ(Filter);
+//REQ(FocusedElementHelp);
+//REQ(CopyToClipboard);
+//REQ(PasteFromClipboard);
+
+cfestring specialkeys::FilterListQuestion(cfestring what)
+{
+  festring What=what;
+
+  v2 pos = v2(16, 6);
+  static festring Topic = "Type this list filter:";
+
+  int R = iosystem::StringQuestion(What, Topic, pos, WHITE, 0, 30, false /*TODO !bGameIsRunning*/, true, NULL);
+
+  /**
+   * TODO
+   * clear the filter text using the background!
+   * needed when pressing ESC or ENTER and not changing the filter!
+   * tip: igraph::BlitBackGround(pos, v2(RES.X, 23));
+   * also, could may be copy the background from doublebuffer and just paste it back after....
+   */
+
+  // cheap workaround to not look too bad (like input was not accepted) at least
+  //  DOUBLE_BUFFER->Fill(pos,v2(RES.X, 23),BLACK);
+  FONT->Printf(DOUBLE_BUFFER, pos, BLACK, "%s", Topic.CStr());
+  FONT->Printf(DOUBLE_BUFFER, v2(pos.X, pos.Y + 10), BLACK, "%s_", What.CStr());
+
+  if(R == NORMAL_EXIT){ DBG1(What.CStr());
+    return What;
+  }
+
+  if(R == ABORTED)
+    return what;
+
+  return cfestring();
+}
+
+bool specialkeys::IsRequestedEvent(SKEvent e)
+{
+  return (e==Request);
+}
+
+std::vector<festring> afsHelpDialog;
+void DrawHelpDialog(bitmap* Buffer) //TODO this kind'o message should be more global to be easier to be used...
+{ //TODO create a buffer to not re-draw every loop... unless want to animate it...
+  if(afsHelpDialog.size()==0)return;
+
+  int iFontLetterWidth=8;
+  int iLH=15;
+
+  v2 v2Border;
+  v2Border.Y=afsHelpDialog.size()*iLH+iLH*2;
+  for(int i=0;i<afsHelpDialog.size();i++){
+    int iW = afsHelpDialog[i].GetSize()*iFontLetterWidth;
+    if(v2Border.X<iW)v2Border.X=iW;
+  }
+
+  v2Border.X+=iLH*2;
+
+  v2 v2TL(RES.X/2-v2Border.X/2,RES.Y/2-v2Border.Y/2);
+
+  Buffer->Fill(v2TL,v2Border,BLACK);
+  graphics::DrawRectangleOutlineAround(Buffer, v2TL, v2Border, DARK_GRAY, true);
+
+  v2TL+=v2(iLH,iLH);
+
+  for(int i=0;i<afsHelpDialog.size();i++)
+    FONT->Printf(Buffer, v2(v2TL.X,v2TL.Y+i*iLH), WHITE, "%s", afsHelpDialog[i].CStr());
+}
+
+bool bConsumingEvent=false;
+bool specialkeys::IsConsumingEvent()
+{
+  return bConsumingEvent;
+}
+
+int specialkeys::Request=-1;
+bool specialkeys::ConsumeEvent(SKEvent e,festring& fsInOut){DBGLN;
+  if(!IsRequestedEvent(e))
+    return false;
+
+  bConsumingEvent=true;
+
+  switch(e){
+    case ClearStringInput:{
+      fsInOut.Empty();
+      Request=-1;
+    };break;
+
+    case Filter:{DBGLN;
+      globalwindowhandler::SuspendKeyTimeout();
+      fsInOut=FilterListQuestion(fsInOut);
+      globalwindowhandler::ResumeKeyTimeout();
+  //    bFilterRequest=false;
+      Request=-1;
+    };break;
+
+    case CopyToClipboard:{DBGLN;
+      #ifdef UNIX //TODO works for MACOSX too?
+        std::ostringstream osStkCmd;
+        osStkCmd<<"echo -n '"<<fsInOut.CStr()<<"' |xclip -i";
+        FILE* pipeFile = popen(osStkCmd.str().c_str(),"r"); //TODO exec pipelessly
```

Comment:
Safer: It should popen with write mode and write the content of fsInOut to the pipeFile.

Reply:
> MacOS uses pbcopy and pbpaste.
oh, well no way I can test this but is interesting, windows also have a specific implementation I cant test either

Reply:
> Danger: The content of fsInOut is not escaped for the shell string and it could contain dashes -anything or some dangerous symbols and ruin your disks.

Actually, from `help echo`, there is no way it could damage the disk thru a dash command, as fsInOut is actually within `echo -n '[HERE]'`. Or you think there is a way thru a dash option?

But, it could contain a subshell like 
```
$() or ``
```
and inside it some command like $(fdisk) `fdisk`
but still, other than using the disk a lot or filling it up thru some shell script loop
w/o sudo password I see no real danger here
unless I am missing something?

but even still, 
the way `echo` works, using ' and not " will prevent subshells from being executed at all!!

if I had coded `echo "..."` it could allow subshells, but not like it is already `echo '...'`
unless I am missing something again?

> Safer: It should popen "xclip -i" with write mode and write the content of fsInOut to the pipeFile.

I cant see the advantage of this change.
The safer change would be to actually escape all chars, but being inside `echo '...'` I still cant see the advantage of doing it.
unless I am missing something?

Reply:
The input can contain `'`. fputs(cstr, pipe) is much easier without escaping anything.

For your reference:

```c
#include <stdio.h>
#include <stdlib.h>

void die(int status, char* msg)
{
    if (msg != NULL)
        fprintf(stderr, "\n\nError: %s\n\n", msg);

    exit(status);
}

int main(int argc, char** argv)
{
    if (argc > 0) {
        FILE* pipe = NULL;

        if ((pipe = popen("pbcopy", "w")) == NULL)
            die(1, "failed to open a pipe");

        for (int i = 1; i < argc; i++)
            if ((i > 1 && fputs(" ", pipe) < 0) || fputs(argv[i], pipe) < 0)
                die(2, "failed to write to the pipe");

        if (pclose(pipe) < 0)
            die(3, "failed to close the pipe");
    }

    return 0;
}
```

Reply:
`pbcopy` accepts pipe like xclip in `| xclip -i` ?

Reply:
oh... that implementation is specific for MaxOSX right? it wont work with xclip.
I got that copy to clipboard code tip on stackexchange and just adapted it.
the `for loop` to fill pbcopy with data is strange to me (I am not used to it).
the `popen(` I use is a simple OS command that does everything in one shot where the actual text returned thru the pipe (if any) will just be ignored.
I will just comment out that part to prevent problems then, as it could happen thru a command typed thru an item's label/engrave when copying a list item.

Reply:
I think, as soon fputs() sends '\n' or may be at pclose(), it would just execute any command in case of xclip.
by what it looks like, pbcopy is intended to be used as a pipe.
what I coded for linux was a OS command that creates a pipe thru `echo`, and that can really break thru user ctrl+c.

Reply:
The code for osx is the same for linux, you just need to change the shell commands from "xclip -i" to "pbcopy", "xclip -o" to "pbpaste". `fputs("text", pipe)` equals producing input for the command (e.g. `xclip -i`), so you don't have to risk using `echo`.

The simple demo above is a program for copying the commandline arguments to the clipboard. Just change "pbcopy" to "xclip -i" and try it yourself! `program a b c  # overwrite the clipboard content with "a b c"`

Reply:
oh, cool thx, I will try it as soon I can,
to paste from clipboard and copy from the "string question mode" are other useful features I didnt implement but would like to :)